File: uci/uci_controller.py
from PyQt4.QtCore import *
import re
from util.proc import set_lowpriority
import time
from uci.uci_worker import Uci_worker
import time

class Uci_controller(QObject):

    def __init__(self,parent=None):
        super(Uci_controller,self).__init__(parent)
        self.thread = QThread()
        self.uci_worker = Uci_worker()
        self.timer = QTimer()

        self.timer.timeout.connect(self.uci_worker.process_command)
        self.timer.start(50)

        self.connect(self.uci_worker,SIGNAL("bestmove(QString)"),self.on_bestmove)
        self.connect(self.uci_worker,SIGNAL("info(PyQt_PyObject)"),self.on_info)
        self.connect(self.uci_worker,SIGNAL("on_error(QString)"),self.on_error)
        self.connect(self,SIGNAL("new_command(QString)"),self.uci_worker.add_command)
        self.connect(self,SIGNAL("new_fen(QString)"),self.uci_worker.update_fen)

        # needed to keep reference?
        self.foobar = self.uci_worker.process

        self.timer.moveToThread(self.thread)
        self.uci_worker.moveToThread(self.thread)

        self.thread.start(QThread.LowestPriority)
        self.engine_running = False


    def on_error(self,msg):
        pass

    def send_fen(self,fen_string):
        self.emit(SIGNAL("new_fen(QString)"),fen_string)

    def on_bestmove(self,msg):
        self.emit(SIGNAL("bestmove(QString)"),msg)

    def on_info(self,engine_info):
        self.emit(SIGNAL("updateinfo(PyQt_PyObject)"),engine_info)

    def stop_engine(self):
        print("emitting quit")
        self.emit(SIGNAL("new_command(QString)"),"quit")
        #self.foobar.kill()
        #self.foobar.waitForFinished()
        #self.thread.killTimer(1)
        self.engine_running = False

    def soft_stop_engine(self):
        self.emit(SIGNAL("new_command(QString)"),"stop")
        print("emitting soft stop")

        #self.foobar.kill()
        #self.foobar.waitForFinished()
        #self.thread.killTimer(1)
        #self.engine_running = False

    def kill_engine(self):
        #self.emit(SIGNAL("new_command(QString)"),"quit")
        self.foobar.kill()
        self.foobar.waitForFinished(-1)
        self.engine_running = False

    def start_engine(self,path):
        self.emit(SIGNAL("new_command(QString)"),"start_engine?"+path)
        self.engine_running = True
        print("cold start engine")


#    def reset_engine(self,path):
#        self.stop_engine()
#        self.start_engine(path)

    def uci_newgame(self):
        self.emit(SIGNAL("new_command(QString)"),"ucinewgame")

    def uci_send_command(self, command):
        self.emit(SIGNAL("new_command(QString)"),command)

    def uci_send_position(self,uci_string):
        self.emit(SIGNAL("new_command(QString)"),uci_string)

    def uci_ok(self):
        self.emit(SIGNAL("new_command(QString)"),"uci")

    def uci_go_movetime(self,ms):
        self.emit(SIGNAL("new_command(QString)"),"go movetime "+str(ms))

    # works only with stockfish
    def uci_strength(self,level):
        self.emit(SIGNAL("new_command(QString)"),"setoption name Skill Level value "+str(level))

    def uci_go_infinite(self):
        self.emit(SIGNAL("new_command(QString)"),"go infinite")

    def send_engine_options(self, options):
        for (option,val) in options:
            if(option.type == 'spin'):
                self.uci_send_command("setoption name "+option.name+" value "+str(val))
            elif(option.type == 'check'):
                self.uci_send_command("setoption name "+option.name+" value "+str(val).lower())
            else:
                self.uci_send_command("setoption name "+option.name+" value "+val)

    def soft_reset_engine(self, engine):
        #self.stop_engine()
        #self.start_engine(engine.path)
        if(self.engine_running):
            print("soft stopping (resetting) engine")
            self.uci_send_command("stop")
            self.uci_newgame()
            self.uci_ok()
            self.send_engine_options(engine.options)
        else:
            self.hard_reset_engine(engine)
        #self.uci_newgame()

    def reset_engine(self, engine):
        self.stop_engine()
        self.start_engine(engine.path)
        self.uci_ok()
        self.send_engine_options(engine.options)
        self.uci_newgame()

File: uci/uci_controller.cpp
/* Jerry - A Chess Graphical User Interface
 * Copyright (C) 2014-2016 Dominik Klein
 * Copyright (C) 2015-2016 Karl Josef Klein
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


#include "uci_controller.h"
#include "uci_worker.h"
#include <iostream>
#include <QDebug>
#include "model/engine_option.h"

UciController::UciController(QObject *parent) :
   QObject(parent)
{
    this->thread = new QThread();

    this->worker = new UciWorker();
    this->worker->process = new QProcess();
    this->worker->cmd_queue = new QQueue<QString>();
    this->worker->engine_info = new EngineInfo();
    this->worker->process = new QProcess();
    this->worker->go_infinite = false;
    this->worker->current_fen = QString("");

    this->timer = new QTimer();
    QObject::connect(this->timer, &QTimer::timeout, this->worker, &UciWorker::processCommands);
    this->timer->start(40);

    this->timer->moveToThread(this->thread);
    this->worker->moveToThread(this->thread);
    this->worker->process->moveToThread(this->thread);
    this->thread->start(QThread::LowestPriority);

    QObject::connect(this->worker,&UciWorker::bestmove,this,&UciController::onBestmove);
    QObject::connect(this->worker,&UciWorker::error,this,&UciController::onError);
    QObject::connect(this->worker,&UciWorker::info,this,&UciController::onInfo);
    QObject::connect(this->worker,&UciWorker::bestPv,this,&UciController::onBestPv);
    QObject::connect(this->worker,&UciWorker::mateDetected,this,&UciController::onMateDetected);
    QObject::connect(this->worker,&UciWorker::eval,this,&UciController::onEval);
    QObject::connect(this,&UciController::newCommand,this->worker,&UciWorker::addCommand);
    QObject::connect(this,&UciController::newFen,this->worker,&UciWorker::updateFen);



}

void UciController::onError(QString error) {
    std::cerr << error.toStdString() << std::endl;
}

void UciController::onBestmove(QString uci_move) {
    emit(bestmove(uci_move));
}

void UciController::onInfo(QString engine_info) {
    emit(updateInfo(engine_info));
}

void UciController::onBestPv(QString pvs) {
    emit(bestPv(pvs));
}

void UciController::onMateDetected(int mateIn) {
    emit(mateDetected(mateIn));
}

void UciController::onEval(float ev) {
    emit(eval(ev));
}

void UciController::stopEngine() {
    emit(newCommand(QString("quit")));
}

void UciController::startEngine(const QString &path) {
    emit(newCommand(QString("start_engine?").append(path)));
}

void UciController::uciNewgame() {
    emit(newCommand(QString("ucinewgame")));
}

void UciController::uciSendCommand(const QString &command) {
    emit(newCommand(command));
}

void UciController::uciSendPosition(const QString &position) {
    emit(newCommand(position));
}

void UciController::uciOk() {
    emit(newCommand(QString("uci")));
}

void UciController::uciStrength(int level) {
    QString lvl = QString::number(level);
    emit(newCommand(QString("setoption name Skill Level value ").append(lvl)));
}

void UciController::sendEngineOptions(QList<EngineOption*> *optList) {
    for(int i=0;i<optList->size();i++) {
        QString cmd = optList->at(i)->toUciCommandString();
        if(!cmd.isEmpty()) {
            emit(newCommand(cmd));
        }
    }
}

void UciController::uciGoMovetime(int milliseconds) {
    QString ms = QString::number(milliseconds);
    emit(newCommand(QString("go movetime ").append(ms)));
}

void UciController::uciGoInfinite() {
    emit(newCommand(QString("go infinite")));
}

void UciController::uciSendFen(const QString &fen) {
    emit(newFen(fen));
}
